[scalar-field-filtering] Collection create success if an invalid field name provided

**Describe the bug**
 branch: scalar-field-filtering
```
______________________________________________________________________________ TestCreateCollectionInvalid.test_create_collection_invalid_field_name[12-s] _______________________________________________________________________________

self = <test_create_collection.TestCreateCollectionInvalid object at 0x7f5bdc071630>, connect = <milvus.client.stub.Milvus object at 0x7f5bded984e0>, get_invalid_string = '12-s'

    def test_create_collection_invalid_field_name(self, connect, get_invalid_string):
        collection_name = gen_unique_str(collection_id)
        fields = copy.deepcopy(default_fields)
        field_name = get_invalid_string
        field = {"field": field_name, "type": DataType.INT8}
        fields["fields"].append(field)
        with pytest.raises(Exception) as e:
>           connect.create_collection(collection_name, fields)
E           Failed: DID NOT RAISE <class 'Exception'>

```

**Steps/Code to reproduce behavior**
Follow this [guide](http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) to craft a minimal bug report. This helps us reproduce the issue you're having and resolve the issue more quickly.

**Expected behavior**
A clear and concise description of what you expected to happen.

**Environment details**
- Hardware/Softward conditions (OS, CPU, GPU, Memory)
- Method of installation (Docker, or from source)
- Milvus version (v0.3.1, or v0.4.0)
- Milvus configuration (Settings you made in `server_config.yaml`)

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -16,6 +16,7 @@ Please mark all changes in change log and use the issue from GitHub
 -   \#2637 Suit the range of HNSW parameters
 -   \#2649 Search parameter of annoy has conflict with document
 -   \#2693 Collection create success if no dimension value provided
+-   \#2694 Collection create success if an invalid field name provided
 
 ## Feature
 -   \#2319 Redo metadata to support MVCC

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,7 @@ Please mark all changes in change log and use the issue from GitHub
 -   \#2649 Search parameter of annoy has conflict with document
 -   \#2693 Collection create success if no dimension value provided
 -   \#2694 Collection create success if an invalid field name provided
+-   \#2695 The number of fields should be limited
 
 ## Feature
 -   \#2319 Redo metadata to support MVCC

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -18,6 +18,7 @@ Please mark all changes in change log and use the issue from GitHub
 -   \#2693 Collection create success if no dimension value provided
 -   \#2694 Collection create success if an invalid field name provided
 -   \#2695 The number of fields should be limited
+-   \#2696 Check the validity of the parameters of creating collection: segment_size
 
 ## Feature
 -   \#2319 Redo metadata to support MVCC

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,6 +19,7 @@ Please mark all changes in change log and use the issue from GitHub
 -   \#2694 Collection create success if an invalid field name provided
 -   \#2695 The number of fields should be limited
 -   \#2696 Check the validity of the parameters of creating collection: segment_size
+-   \#2697 Index can not be created
 
 ## Feature
 -   \#2319 Redo metadata to support MVCC

--- a/core/src/db/engine/ExecutionEngineImpl.cpp
+++ b/core/src/db/engine/ExecutionEngineImpl.cpp
@@ -958,6 +958,7 @@ ExecutionEngineImpl::HybridSearch(scheduler::SearchJobPtr search_job,
     search_job->SetVectors(vectors);
     search_job->vector_count() = nq;
     search_job->topk() = topk;
+    search_job->vector_params() = vector_query->extra_params;
 
     status = Search(search_ids, distances, search_job, hybrid);
     if (!status.ok()) {

--- a/core/src/scheduler/job/SearchJob.h
+++ b/core/src/scheduler/job/SearchJob.h
@@ -136,6 +136,11 @@ class SearchJob : public Job {
         return vector_count_;
     }
 
+    milvus::json&
+    vector_params() {
+        return extra_params_;
+    }
+
     SearchTimeStat&
     time_stat() {
         return time_stat_;

--- a/core/src/scheduler/task/SearchTask.cpp
+++ b/core/src/scheduler/task/SearchTask.cpp
@@ -123,21 +123,6 @@ XSearchTask::XSearchTask(const std::shared_ptr<server::Context>& context, Segmen
         if (!file_->index_params_.empty()) {
             json_params = milvus::json::parse(file_->index_params_);
         }
-        //        if (auto job = job_.lock()) {
-        //            auto search_job = std::static_pointer_cast<scheduler::SearchJob>(job);
-        //            query::GeneralQueryPtr general_query = search_job->general_query();
-        //            if (general_query != nullptr) {
-        //                std::unordered_map<std::string, engine::DataType> types;
-        //                auto attr_type = search_job->attr_type();
-        //                auto type_it = attr_type.begin();
-        //                for (; type_it != attr_type.end(); type_it++) {
-        //                    types.insert(std::make_pair(type_it->first, (engine::DataType)(type_it->second)));
-        //                }
-        //                index_engine_ =
-        //                    EngineFactory::Build(file_->dimension_, file_->location_, engine_type,
-        //                                         (MetricType)file_->metric_type_, types, json_params);
-        //            }
-        //        }
         index_engine_ = EngineFactory::Build(file_->dimension_, file_->location_, engine_type,
                                              (MetricType)file_->metric_type_, json_params);
     }

--- a/core/src/server/ValidationUtil.cpp
+++ b/core/src/server/ValidationUtil.cpp
@@ -30,6 +30,7 @@ constexpr int64_t COLLECTION_DIMENSION_LIMIT = 32768;
 constexpr int32_t INDEX_FILE_SIZE_LIMIT = 4096;  // index trigger size max = 4096 MB
 constexpr int64_t M_BYTE = 1024 * 1024;
 constexpr int64_t MAX_INSERT_DATA_SIZE = 256 * M_BYTE;
+constexpr int64_t FIELD_NAME_SIZE_LIMIT = 255;
 
 Status
 CheckParameterRange(const milvus::json& json_params, const std::string& param_name, int64_t min, int64_t max,
@@ -122,6 +123,44 @@ ValidateCollectionName(const std::string& collection_name) {
             std::string msg = invalid_msg + "Collection name can only contain numbers, letters, and underscores.";
             LOG_SERVER_ERROR_ << msg;
             return Status(SERVER_INVALID_COLLECTION_NAME, msg);
+        }
+    }
+
+    return Status::OK();
+}
+
+Status
+ValidateFieldName(const std::string& field_name) {
+    // Field name shouldn't be empty.
+    if (field_name.empty()) {
+        std::string msg = "Field name should not be empty.";
+        LOG_SERVER_ERROR_ << msg;
+        return Status(SERVER_INVALID_FIELD_NAME, msg);
+    }
+
+    std::string invalid_msg = "Invalid field name: " + field_name + ". ";
+    // Field name size shouldn't exceed 255.
+    if (field_name.size() > FIELD_NAME_SIZE_LIMIT) {
+        std::string msg = invalid_msg + "The length of a field name must be less than 255 characters.";
+        LOG_SERVER_ERROR_ << msg;
+        return Status(SERVER_INVALID_FIELD_NAME, msg);
+    }
+
+    // Field name first character should be underscore or character.
+    char first_char = field_name[0];
+    if (first_char != '_' && std::isalpha(first_char) == 0) {
+        std::string msg = invalid_msg + "The first character of a field name must be an underscore or letter.";
+        LOG_SERVER_ERROR_ << msg;
+        return Status(SERVER_INVALID_FIELD_NAME, msg);
+    }
+
+    int64_t field_name_size = field_name.size();
+    for (int64_t i = 1; i < field_name_size; ++i) {
+        char name_char = field_name[i];
+        if (name_char != '_' && std::isalnum(name_char) == 0) {
+            std::string msg = invalid_msg + "Field name cannot only contain numbers, letters, and underscores.";
+            LOG_SERVER_ERROR_ << msg;
+            return Status(SERVER_INVALID_FIELD_NAME, msg);
         }
     }
 

--- a/core/src/server/ValidationUtil.h
+++ b/core/src/server/ValidationUtil.h
@@ -28,6 +28,9 @@ extern Status
 ValidateCollectionName(const std::string& collection_name);
 
 extern Status
+ValidateFieldName(const std::string& field_name);
+
+extern Status
 ValidateTableDimension(int64_t dimension, int64_t metric_type);
 
 extern Status

--- a/core/src/server/delivery/hybrid_request/CreateHybridCollectionRequest.cpp
+++ b/core/src/server/delivery/hybrid_request/CreateHybridCollectionRequest.cpp
@@ -108,7 +108,12 @@ CreateHybridCollectionRequest::OnExecute() {
         collection_info.collection_id_ = collection_name_;
         collection_info.dimension_ = dimension;
         if (extra_params_.contains("segment_size")) {
-            collection_info.index_file_size_ = extra_params_["segment_size"].get<int64_t>();
+            auto segment_size = extra_params_["segment_size"].get<int64_t>();
+            collection_info.index_file_size_ = segment_size;
+            status = ValidateCollectionIndexFileSize(segment_size);
+            if (!status.ok()) {
+                return status;
+            }
         }
 
         if (vector_param.contains("metric_type")) {

--- a/core/src/server/delivery/hybrid_request/CreateHybridCollectionRequest.cpp
+++ b/core/src/server/delivery/hybrid_request/CreateHybridCollectionRequest.cpp
@@ -75,6 +75,11 @@ CreateHybridCollectionRequest::OnExecute() {
         for (auto& field_type : field_types_) {
             engine::meta::hybrid::FieldSchema schema;
             auto field_name = field_type.first;
+            status = ValidateFieldName(field_name);
+            if (!status.ok()) {
+                return status;
+            }
+
             auto index_params = field_index_params_.at(field_name);
             schema.collection_id_ = collection_name_;
             schema.field_name_ = field_name;

--- a/core/src/server/grpc_impl/GrpcRequestHandler.cpp
+++ b/core/src/server/grpc_impl/GrpcRequestHandler.cpp
@@ -656,7 +656,7 @@ GrpcRequestHandler::DropCollection(::grpc::ServerContext* context, const ::milvu
 ::grpc::Status
 GrpcRequestHandler::CreateIndex(::grpc::ServerContext* context, const ::milvus::grpc::IndexParam* request,
                                 ::milvus::grpc::Status* response) {
-    CHECK_NULLPTR_RETURN(request);
+    CHECK_NULLPTR_RETURN(request)
     LOG_SERVER_INFO_ << LogOut("Request [%s] %s begin.", GetContext(context)->RequestID().c_str(), __func__);
 
     milvus::json json_params;
@@ -666,6 +666,8 @@ GrpcRequestHandler::CreateIndex(::grpc::ServerContext* context, const ::milvus::
             json_params = json::parse(extra.value());
         }
     }
+
+    std::string param = json_params.dump();
 
     Status status = request_handler_.CreateIndex(GetContext(context), request->collection_name(), request->field_name(),
                                                  request->index_name(), json_params);

--- a/core/src/server/grpc_impl/GrpcRequestHandler.cpp
+++ b/core/src/server/grpc_impl/GrpcRequestHandler.cpp
@@ -33,6 +33,7 @@ namespace server {
 namespace grpc {
 
 const char* EXTRA_PARAM_KEY = "params";
+const size_t MAXIMUM_FIELD_NUM = 64;
 
 ::milvus::grpc::ErrorCode
 ErrorMap(ErrorCode code) {
@@ -51,6 +52,8 @@ ErrorMap(ErrorCode code) {
         {SERVER_INVALID_COLLECTION_NAME, ::milvus::grpc::ErrorCode::ILLEGAL_COLLECTION_NAME},
         {SERVER_INVALID_COLLECTION_DIMENSION, ::milvus::grpc::ErrorCode::ILLEGAL_DIMENSION},
         {SERVER_INVALID_VECTOR_DIMENSION, ::milvus::grpc::ErrorCode::ILLEGAL_DIMENSION},
+        {SERVER_INVALID_FIELD_NAME, ::milvus::grpc::ErrorCode::ILLEGAL_ARGUMENT},
+        {SERVER_INVALID_FIELD_NUM, ::milvus::grpc::ErrorCode::ILLEGAL_ARGUMENT},
 
         {SERVER_INVALID_INDEX_TYPE, ::milvus::grpc::ErrorCode::ILLEGAL_INDEX_TYPE},
         {SERVER_INVALID_ROWRECORD, ::milvus::grpc::ErrorCode::ILLEGAL_ROWRECORD},
@@ -576,6 +579,12 @@ GrpcRequestHandler::CreateCollection(::grpc::ServerContext* context, const ::mil
     std::unordered_map<std::string, engine::meta::hybrid::DataType> field_types;
     std::unordered_map<std::string, milvus::json> field_index_params;
     std::unordered_map<std::string, std::string> field_params;
+    if (request->fields_size() > MAXIMUM_FIELD_NUM) {
+        Status status = Status{SERVER_INVALID_FIELD_NUM, "Maximum field's number should be limited to 64"};
+        LOG_SERVER_INFO_ << LogOut("Request [%s] %s end.", GetContext(context)->RequestID().c_str(), __func__);
+        SET_RESPONSE(response, status, context);
+        return ::grpc::Status::OK;
+    }
     for (int i = 0; i < request->fields_size(); ++i) {
         auto field = request->fields(i);
         auto field_name = field.name();

--- a/core/src/utils/Error.h
+++ b/core/src/utils/Error.h
@@ -93,6 +93,7 @@ constexpr ErrorCode SERVER_OUT_OF_MEMORY = ToServerErrorCode(117);
 constexpr ErrorCode SERVER_INVALID_PARTITION_TAG = ToServerErrorCode(118);
 constexpr ErrorCode SERVER_INVALID_BINARY_QUERY = ToServerErrorCode(119);
 constexpr ErrorCode SERVER_INVALID_DSL_PARAMETER = ToServerErrorCode(120);
+constexpr ErrorCode SERVER_INVALID_FIELD_NAME = ToServerErrorCode(121);
 
 // db error code
 constexpr ErrorCode DB_META_TRANSACTION_FAILED = ToDbErrorCode(1);

--- a/core/src/utils/Error.h
+++ b/core/src/utils/Error.h
@@ -94,6 +94,7 @@ constexpr ErrorCode SERVER_INVALID_PARTITION_TAG = ToServerErrorCode(118);
 constexpr ErrorCode SERVER_INVALID_BINARY_QUERY = ToServerErrorCode(119);
 constexpr ErrorCode SERVER_INVALID_DSL_PARAMETER = ToServerErrorCode(120);
 constexpr ErrorCode SERVER_INVALID_FIELD_NAME = ToServerErrorCode(121);
+constexpr ErrorCode SERVER_INVALID_FIELD_NUM = ToServerErrorCode(122);
 
 // db error code
 constexpr ErrorCode DB_META_TRANSACTION_FAILED = ToDbErrorCode(1);

--- a/sdk/examples/simple/src/ClientTest.cpp
+++ b/sdk/examples/simple/src/ClientTest.cpp
@@ -278,10 +278,10 @@ ClientTest::SearchEntitiesByID(const std::string& collection_name, int64_t topk,
 }
 
 void
-ClientTest::CreateIndex(const std::string& collection_name, milvus::IndexType type, int64_t nlist) {
+ClientTest::CreateIndex(const std::string& collection_name, int64_t nlist) {
     milvus_sdk::TimeRecorder rc("Create index");
     std::cout << "Wait until create all index done" << std::endl;
-    JSON json_params = {{"nlist", nlist}};
+    JSON json_params = {{"nlist", nlist}, {"index_type", "IVFFLAT"}};
     milvus::IndexParam index1 = {collection_name, "field_3", "index_3", json_params.dump()};
     milvus_sdk::Utils::PrintIndexParam(index1);
     milvus::Status stat = conn_->CreateIndex(index1);
@@ -357,12 +357,11 @@ ClientTest::Test() {
     Flush(collection_name);
     GetCollectionStats(collection_name);
 
-    //    BuildSearchEntities(NQ, dim);
     GetEntityByID(collection_name, search_id_array_);
     SearchEntities(collection_name, TOP_K, NPROBE);
-    //    SearchEntitiesByID(collection_name, TOP_K, NPROBE);
 
-    //    CreateIndex(collection_name, INDEX_TYPE, NLIST);
+    CreateIndex(collection_name, NLIST);
+    SearchEntities(collection_name, TOP_K, NPROBE);
     GetCollectionStats(collection_name);
 
     std::vector<int64_t> delete_ids = {search_id_array_[0], search_id_array_[1]};

--- a/sdk/examples/simple/src/ClientTest.h
+++ b/sdk/examples/simple/src/ClientTest.h
@@ -63,7 +63,7 @@ class ClientTest {
     SearchEntitiesByID(const std::string&, int64_t, int64_t);
 
     void
-    CreateIndex(const std::string&, milvus::IndexType, int64_t);
+    CreateIndex(const std::string&, int64_t);
 
     void
     LoadCollection(const std::string&);

